### ISSUE TITLE
Pod is stuck in PodInitializing state after adding the inject annotation

### ISSUE THREAD

**Issue by AAber** [linkerd/linkerd2#7179]
### What is the issue?

After I added the inject annotation to my deployment the pods are stuck in PodInitializing state

### How can it be reproduced?

Add the following line to your deployment yaml and redeploy:
`dnsPolicy: None`

### Logs, error output, etc

No need, I bypassed the issue 

#### `linkerd check` output

All is good

### Environment

- Kubernetes Version: 1.20
- Cluster Environment: EKS
- Host OS: Amazon Linux
- Linkerd version: stable-2.11.1

### Possible solution

Remove the following configuration from your deployment yaml:
`dnsPolicy: None`

### Additional context

I don't know if this is a bug or not but I hope this post will save someone else the time it took me to figure out what the issue is.

**Comment by olix0r** [comment]
@alpeb is correct: the proxy needs to resolve control plane components during startup to obtain mTLS certificates and perform service/policy/etc discovery. If the pod's DNS configuration does not include a DNS server capable of resolving Kubernetes service SRV records, the proxy won't be able to start and, by default, this will block other containers from being initialized.

Thanks for sharing this detail, @AAber. When you were going through this, is there anything that Linkerd could have done differently to make this easier to understand?

**Comment by AAber** [comment]
Just to clarify the setup that caused the issue:
The deployment I was testing is running in production and had latency issues due to DNS resolving issues.
To resolve the DNS latency I configured the pod to use the local VPC DNS server:
(The second IP in your VPC subnet x.x.x.2 ) 
The full DNS config of the pod that had the issue.
``      dnsPolicy: None ``
``      dnsConfig:``
  ``          nameservers:``
  ``            - 10.0.0.2``

So the pod had proper DNS resolving, but no K8s internal resolving!

I think it would help if one of the Linkerd pod's logs had reported a DNS resolving errors.
I looked at the output of "kubectl get events -n linkerd" if the resolving issue was reported in the events log of the linkerd namespace it would be clear that the DNS issue needs to be corrected.

Thanks.
